### PR TITLE
Spec: Contributions, an alternative to CredGraph

### PR DESCRIPTION
**Pull request by blueridger**
# Description
Here's a sketch for a simple arithmetic-based scoring algorithm that could provide an alternative to pagerank. It does not use a cyclic graph, and instead uses a **list of contributions** that have **equations** for weight combination, acting more like a tree data structure with extensibility to act like a directed acyclic graph. The equation type is recursive, and can be used to represent rather complex weight equations such as the example discord message equation:
```
channelWeight * ((emojiWeight1 * roleWeight1) + ... + (emojiWeightN * roleWeightN))
```

# Benefits:

### Memory efficient / scalable
Contributions can be scored one by one, making it much more memory efficient to do all scoring in memory, and also opening up additional memory optimization opportunities involving simple read and write streams, and partial re-scores.

### Intuitively Configure Weights
Configuring weights is one of the large points of friction for setting up a new instance, because of the ambiguity on how weights will change the results. With this model, plugins will have consistent equations that make it easy to understand how a weight change will be factored into the scores.

### Improved In-memory Simulation
This allows more complex weighting logic to be included in the browser-friendly scoring step, meaning that we can shift most of the configuration out of the disk-based "graphing" step. This will allow us to offer AWESOME simulation tools that makes it super easy for end-users to run their own simulations and propose changes to most of the important configurations.

### Contribution-level analytics / End-user accessibility
Right now, it is super difficult to reason about where a participant's cred score is coming from and about where a contribution's score is coming from. This let's our analytic tools go beyond "how much cred does each person have," letting end-users easily audit the end-to-end flow of cred through contributions.

### Composability / Potential for more complex inter-plugin and inter-instance logic
Right now, inter-plugin logic is difficult and wonky, and inter-instance logic is pretty much out of the question. This new model will allow us to use similar "equation" configuration to add additional layers of score aggregation across plugins, and even across instances!

### Accessibility for Devs (1st and 3rd party)
It is difficult to onboard to both our core code and our 3rd-party library because both generally require intimate knowledge of the CredRank algorithm. This new model will make it SUPER easy for new devs to onboard to our team or start building advanced integrations.

### Powerful plugin creation out-of-the-box
The discord plugin is our most nuanced, but only because it has a bunch of hacky logic that breaks the intended architecture. This new model prioritizes and creates structure for embedding the nuance that has proven to be so desirable (as shown by the constant requests for additional weight configuration types on categories, tags, repos, etc). This means that advanced plugin creation will be more accessible and faster. This model additionally has GREAT potential for developers to define a schema that our utilities then use to generate builders, parsers, validators, declarations, and default configurations, further optimizing plugin creation.

### Extensible
I only included support for ADD and MULTIPLY but this could easily be extended to include operations such as MAX, MIN, OR_ELSE, AVERAGE. And any such extensions would then be immediately available for all plugins to integrate!!!

### Can be supported in parallel with CredRank
We can create two different CLI paths for CredRank and Contributions, allowing some instances to keep using CredRank while others experiment with this new model. If the new model is as successful as I suspect, we can the discuss deprecating CredRank.

### It's basically done
This PR includes most of the heavy lifting / design required. The only things left would be some data piping through new CLI commands, and converting each plugin (one at a time) to support the new scoring path.

# Downsides:
Honestly, I'm having a hard time thinking of any. Maybe there are some future analytics that we haven't built yet (and have no plans to build) that would benefit from a graph model? Maybe people understanding the algorithm better will mean gamers understanding how to game the algorithm better?

# Launch plan
A core algorithm change is a big deal, so we will be careful/ conservative in how we introduce it. This will initially be launched as its own CLI, an optional alternative to using the graph/credrank CLI. We will find some instances interested in converting to the new algorithm, or run parallel test instances, and see how it performs and is received by those beta testers. If it performs very well, we may consider discontinuing support for CredRank. This will be an ecosystem-wide discussion and we will strive for inclusivity in that choice.

# Test Plan
1. Remove all plugins except discord from sourcecred.json
1. `scdev graph && scdev credrank`

# Test Plan for 1st commit
from `packages/sourcecred` run `yarn shell`
```
$ sc.test.score(sc.test.exampleDiscordContribution, sc.test.exampleDiscordWeightConfig)
{
  id: '1234',
  plugin: 'Discord',
  type: 'Message',
  participants: [
    { id: 'ASDVERW342tFD', shares: 1 },
    { id: 'SD533FEfdsdrG', shares: 3 }
  ],
  equation: {
    type: 'MULTIPLY',
    description: 'contribution attributes',
    factors: [ [Object] ],
    equationFactors: [ [Object] ]
  },
  score: 6
}
```

### CODE REVIEW

**Comment by blueridger** [comment]
Thena | 23434.3 | 8.89% | 101275.46 | 11.75% | 2.86%
-- | -- | -- | -- | -- | --
hz | 6960.6 | 2.64% | 32676.74 | 3.79% | 1.15%
scrabbleboy | 6853.7 | 2.60% | 31003.67 | 3.60% | 1.00%
topocount | 11981.3 | 4.54% | 47493.24 | 5.51% | 0.96%
Ryeder | 22430.2 | 8.51% | 81404.51 | 9.44% | 0.93%
Willow | 3068.4 | 1.16% | 14599.93 | 1.69% | 0.53%
echojuliet | 6850.1 | 2.60% | 26080.44 | 3.03% | 0.43%
magwalk | 6640.5 | 2.52% | 24903.17 | 2.89% | 0.37%
finitechris | 1988.6 | 0.75% | 9512.56 | 1.10% | 0.35%
AL0YSI0US | 11578.6 | 4.39% | 40779.71 | 4.73% | 0.34%
saintmedusa | 6957.6 | 2.64% | 25425.9 | 2.95% | 0.31%
BrianBelhumeur | 2657.6 | 1.01% | 11308.99 | 1.31% | 0.30%
ezrau | 4805.3 | 1.82% | 18011.32 | 2.09% | 0.27%
Marcie | 2583.8 | 0.98% | 10637.39 | 1.23% | 0.25%
lotusleaf | 4110.3 | 1.56% | 15072.84 | 1.75% | 0.19%
haroldiii | 2693.2 | 1.02% | 10416.87 | 1.21% | 0.19%
Kokomo | 1342.9 | 0.51% | 5808.46 | 0.67% | 0.16%
bsodenkamp | 2426.8 | 0.92% | 9199.41 | 1.07% | 0.15%
leah-v | 228.3 | 0.09% | 1562.6 | 0.18% | 0.09%
Nav89 | 918.2 | 0.35% | 3719.89 | 0.43% | 0.08%
Magey | 1613.3 | 0.61% | 5958.93 | 0.69% | 0.08%
Zizi | 563.9 | 0.21% | 2434.21 | 0.28% | 0.07%
Threm | 335.5 | 0.13% | 1665.39 | 0.19% | 0.07%
amrro | 1153.6 | 0.44% | 4273.35 | 0.50% | 0.06%
prose11 | 709.7 | 0.27% | 2654.7 | 0.31% | 0.04%
Randall | 720.3 | 0.27% | 2609.34 | 0.30% | 0.03%
mahmoud | 540.6 | 0.21% | 1953.54 | 0.23% | 0.02%
daniel-o | 118.4 | 0.04% | 568.9 | 0.07% | 0.02%
TheCryptoCPA | 317.1 | 0.12% | 1185.93 | 0.14% | 0.02%
Dana | 255.4 | 0.10% | 925.46 | 0.11% | 0.01%
Gnomeski | 210.2 | 0.08% | 704.92 | 0.08% | 0.00%
meeting-attendance | 182.6 | 0.07% | 608.63 | 0.07% | 0.00%
JavierCanovas | 244.4 | 0.09% | 778 | 0.09% | 0.00%
Taurean | 235.4 | 0.09% | 738.95 | 0.09% | 0.00%
bendertherobot | 103.1 | 0.04% | 305.42 | 0.04% | 0.00%
kstone | 177.9 | 0.07% | 531.13 | 0.06% | -0.01%
Louisgrx | 133 | 0.05% | 383.02 | 0.04% | -0.01%
hospitaltapes | 138.3 | 0.05% | 391.75 | 0.05% | -0.01%
Joshua | 110.5 | 0.04% | 295.73 | 0.03% | -0.01%
bfa | 112.9 | 0.04% | 302.1 | 0.04% | -0.01%
durgadas | 125.5 | 0.05% | 325.16 | 0.04% | -0.01%
dysbulic | 318.3 | 0.12% | 943.66 | 0.11% | -0.01%
EbonyBelle | 999.5 | 0.38% | 3155.07 | 0.37% | -0.01%
The-Dabber-Gentleman | 284.1 | 0.11% | 800.32 | 0.09% | -0.01%
Tani | 113 | 0.04% | 216.5 | 0.03% | -0.02%
merrick | 125.4 | 0.05% | 245.95 | 0.03% | -0.02%
ER | 109.3 | 0.04% | 189.37 | 0.02% | -0.02%
FredDog | 1028.7 | 0.39% | 3178.62 | 0.37% | -0.02%
rachel-horn | 141.9 | 0.05% | 270 | 0.03% | -0.02%
Hernando-Guzm-n | 442 | 0.17% | 1224.25 | 0.14% | -0.03%
Juankbell | 282.6 | 0.11% | 684 | 0.08% | -0.03%
Lauren-Carl | 165.1 | 0.06% | 292.35 | 0.03% | -0.03%
Mr--Rock | 133.8 | 0.05% | 148.43 | 0.02% | -0.03%
--ecret-James | 182.6 | 0.07% | 305.94 | 0.04% | -0.03%
Felix | 847.5 | 0.32% | 2453.72 | 0.28% | -0.04%
mozilla | 206.8 | 0.08% | 337.5 | 0.04% | -0.04%
tua | 801.9 | 0.30% | 2242.87 | 0.26% | -0.04%
doctorrobinson | 238.8 | 0.09% | 363.5 | 0.04% | -0.05%
pablomendez-95 | 190.7 | 0.07% | 188 | 0.02% | -0.05%
brianlitwin | 298.8 | 0.11% | 537 | 0.06% | -0.05%
Kris | 195.8 | 0.07% | 193.52 | 0.02% | -0.05%
panchomiguel | 12444.9 | 4.72% | 40242.38 | 4.67% | -0.05%
amico | 502 | 0.19% | 1094.45 | 0.13% | -0.06%
lanski | 287.4 | 0.11% | 327 | 0.04% | -0.07%
Nxo | 505 | 0.19% | 974.5 | 0.11% | -0.08%
kus | 303.2 | 0.12% | 288.56 | 0.03% | -0.08%
coopahtroopa | 735.5 | 0.28% | 1331.34 | 0.15% | -0.12%
geleeroyale | 448.8 | 0.17% | 374 | 0.04% | -0.13%
burrrata | 588.8 | 0.22% | 778.13 | 0.09% | -0.13%
Simple-Poll | 423.1 | 0.16% | 167 | 0.02% | -0.14%
MEE6 | 526.8 | 0.20% | 355 | 0.04% | -0.16%
crisog | 752.9 | 0.29% | 1078.83 | 0.13% | -0.16%
cortanav | 904.5 | 0.34% | 1434.69 | 0.17% | -0.18%
evan | 923.7 | 0.35% | 1187.02 | 0.14% | -0.21%
mzargham | 1158.9 | 0.44% | 1876.81 | 0.22% | -0.22%
beanow | 1766.4 | 0.67% | 3847.36 | 0.45% | -0.22%
eeli | 1747.9 | 0.66% | 3765.94 | 0.44% | -0.23%
s-ben | 19822.3 | 7.52% | 62483.71 | 7.25% | -0.27%
kate-huntoon | 1155 | 0.44% | 1333.35 | 0.15% | -0.28%
benoxmo | 2537.4 | 0.96% | 5575.4 | 0.65% | -0.32%
peth | 1243.8 | 0.47% | 1216 | 0.14% | -0.33%
youngkidwarrior | 1642.2 | 0.62% | 2499.05 | 0.29% | -0.33%
wchargin | 2973.4 | 1.13% | 6429.48 | 0.75% | -0.38%
lbstrobbe | 17066.3 | 6.47% | 52193.01 | 6.05% | -0.42%
joiecousins | 9478.6 | 3.60% | 26456.36 | 3.07% | -0.53%
bex | 5787.1 | 2.20% | 10796.94 | 1.25% | -0.94%
hammad | 9299.8 | 3.53% | 19600.89 | 2.27% | -1.25%
KuraFire | 14492.5 | 5.50% | 34095.21 | 3.96% | -1.54%
sandpiper | 8391.5 | 3.18% | 11789.53 | 1.37% | -1.82%

**Comment by topocount** [comment]
Really exicted about the exploration of preserving graph semantics, as discussed in the sprint planning meeting today.

File: packages/sourcecred/src/api/plugin.js [hunk 1]
```diff
@@ -22,9 +22,210 @@ export interface Plugin {
...
+          );
+        },
+        0
+      );
+      factorResult = equation.factors.reduce((accumulator, factor) => {
+        const config = weightConfig.find((x) => x.key === factor.key);
+        if (config === undefined) throw "Unexpected factor key " + factor.key;
+        return (
+          (config.values.find((x) => x.value === factor.value)?.weight ||
+            config.default) + (accumulator === undefined ? 1 : accumulator)
```
Comment (dysbulic) [inline packages/sourcecred/src/api/plugin.js:178]:
```suggestion
          (config.values.find((x) => x.value === factor.value)?.weight ??
            config.default) + (accumulator === undefined ? 1 : accumulator)
```

When `value.weight === 0`, won't it use the default value?

Comment (blueridger) [reply packages/sourcecred/src/api/plugin.js:178]: ah good catch

File: packages/sourcecred/src/api/plugin.js [hunk 2]
```diff
@@ -22,9 +22,210 @@ export interface Plugin {
...
+            (accumulator === undefined ? 1 : accumulator)
+          );
+        },
+        1
+      );
+      factorResult = equation.factors.reduce((accumulator, factor) => {
+        const config = weightConfig.find((x) => x.key === factor.key);
+        if (config === undefined) throw "Unexpected factor key " + factor.key;
+        return (
+          (config.values.find((x) => x.value === factor.value)?.weight ||
```
Comment (blueridger) [inline packages/sourcecred/src/api/plugin.js:152]:
```suggestion
          (config.values.find((x) => x.value === factor.value)?.weight ??
```

**Comment by blueridger** [comment]
Thought of another benefit: Squashing and archiving historical cred will be very easy. This would be a huge scalability win.

**Comment by AL0YSI0US** [comment]
> Do you think credrank would still be applicable to the larger meta-graph across many instances?
> 
> I'm also wondering if/how this could help deal with the issue of cred inflation, which would be more of an issue when combining multiple graphs across instances. How do we ensure that if one sub-graph 10x its cred scores it doesnt affect the higher level graph allocation, instead just preserving the relative ratios. - @META-DREAMER 

I enjoyed seeing this question here. I've peeked under the hood of some instances making use of multipliers that could do just that. I too am looking forward to a global representation of cred for users creating impact across the ecosystem.  

>

**Comment by blueridger** [comment]
@META-DREAMER yes, that isn't included in the MVP, but that sort of anti-inflation logic could very easily be added as a new operator type, and collections of contributions (by plugin or team or dao) can easily be abstracted into contributions by new "meta plugins" that compose ecosystem-level cred scores. Of course, if meta plugins start recursively capturing each other, then cyclic graphs are introduced and will need a mechanism designed to prevent infinite cred loops. We can cross that bridge when we get there, though.

**Pull request closed without merging**